compiler/src/iree/compiler/Utils/unittests/UtilsTest.cpp

```diff
@@ -19,3 +22,47 @@ TEST(Permutation, MakeMovePermutation) {
   EXPECT_THAT(makeMovePermutation(3, 1, 2), ElementsAre(0, 2, 1));
   EXPECT_THAT(makeMovePermutation(3, 2, 0), ElementsAre(2, 0, 1));
 }
+
+TEST(EmbeddedDataDirectory, AddFileGetFile) {
+  EmbeddedDataDirectory dir;
+  EXPECT_TRUE(dir.addFile("filename1", "file contents 1"));
+  EXPECT_TRUE(dir.addFile("filename2", "file contents 2"));
+  EXPECT_FALSE(dir.addFile("filename1", "file contents 3"));
+  EXPECT_EQ(dir.getFile("filename1"), "file contents 1");
+  EXPECT_EQ(dir.getFile("filename2"), "file contents 2");
+  EXPECT_EQ(dir.getFile("filename3"), std::nullopt);
+}
+
+TEST(EmbeddedDataDirectory, WithGlobal) {
+  std::vector<std::thread> threads;
+  for (int i = 0; i < 3; ++i) {
+    threads.emplace_back([i] {
+      EmbeddedDataDirectory::withGlobal([i](EmbeddedDataDirectory &globalDir) {
+        EXPECT_TRUE(globalDir.addFile(llvm::formatv("filename{0}", i).str(),
```

Comment:
For simple string concatenation, `(llvm::Twine("filename") + i).str()` should have less overhead (no need to parse the format str.

Reply:
The format str parsing should happen at compile-time, no? I've just assumed that, up to details, `llvm::formatv` was similar to c++20 `std::format` or fmt.dev 's `fmt::format`.  These things have more or less equivalent interfaces, and at least `fmt::format` boasts compile-time format generation.

Reply:
No, `llvm::formatv` parses the format str at runtime

Reply:
There's nothing in its design/implementation that makes it evaluated at compilation time. I know it won't matter much in unit tests, that's why it's a 'nit'.

Reply:
AFAIK, the generally accepted way to concatenate simple strings / integers / etc. in llvm is with `Twine`